File: typescript/singly_linked_list.ts
type TNode<T> = {
    data: T;
    next?: TNode<T>;
};

interface ILinkedList<T> {
    get length(): number;
    insertAt(data: T, idx: number): void;
    remove(data: T): T | undefined;
    removeAt(idx: number): T | undefined;
    append(data: T): void;
    prepend(data: T): void;
    get(idx: number): T | undefined;
    print(): void;
}

class LinkedList<T> implements ILinkedList<T> {
    private head: TNode<T> | undefined;
    private tail: TNode<T> | undefined;
    public length: number;

    constructor() {
        this.head = this.tail = undefined;
        this.length = 0;
    }

    insertAt(data: T, idx: number): void {
        if (idx < 0 || idx > this.length) {
            throw new Error("index out of bounds");
        }

        if (idx === 0) {
            this.prepend(data);
            return;
        }

        if (idx === this.length) {
            this.append(data);
            return;
        }

        const node = { data: data } as TNode<T>;

        if (this.length === 0) {
            this.head = this.tail = node;
            this.length++;
            return;
        }

        this.length++;

        let previous_node: TNode<T> | undefined = undefined;
        let current_node: TNode<T> | undefined = this.head;

        for (let i = 0; i < idx && current_node; ++i) {
            previous_node = current_node;
            current_node = current_node.next;
        }

        previous_node!.next = node;
        node.next = current_node;
    }

    append(data: T): void {
        const node = { data: data } as TNode<T>;

        if (this.length === 0) {
            this.head = this.tail = node;
            this.length++;
            return;
        }

        this.length++;

        this.tail!.next = node;
        this.tail = node;
    }

    prepend(data: T): void {
        const node = { data: data } as TNode<T>;

        if (this.length === 0) {
            this.head = this.tail = node;
            this.length++;
            return;
        }

        this.length++;

        node.next = this.head;
        this.head = node;
    }

    remove(data: T): T | undefined {
        if (this.length === 0) {
            return;
        }

        if (this.head && this.head.data === data) {
            this.length = Math.max(0, --this.length);

            const rm_node = this.head;

            if (this.length === 0) {
                this.head = this.tail = undefined;

                return rm_node.data;
            }

            this.head = rm_node.next;

            rm_node.next = undefined;

            return rm_node.data;
        }

        let previous_node: TNode<T> | undefined = undefined;
        let current_node: TNode<T> | undefined = this.head;

        while (current_node) {
            if (current_node.data === data) {
                this.length = Math.max(0, --this.length);

                if (this.length === 0) {
                    const rm_node = current_node;

                    this.head = this.tail = undefined;

                    return rm_node.data;
                }

                if (!current_node.next) {
                    const rm_node = current_node;

                    previous_node!.next = undefined;

                    this.tail = previous_node;

                    rm_node.next = undefined;

                    return rm_node.data;
                }

                const rm_node = current_node;

                previous_node!.next = current_node.next;

                rm_node.next = undefined;

                return rm_node.data;
            }

            previous_node = current_node;
            current_node = current_node.next;
        }

        return;
    }

    removeAt(idx: number): T | undefined {
        if (this.length === 0) {
            return;
        }

        if (idx < 0 || idx > this.length) {
            throw new Error("index out of bounds");
        }

        if (this.head && idx === 0) {
            this.length = Math.max(0, --this.length);

            const rm_node = this.head;

            if (this.length === 0) {
                this.head = this.tail = undefined;

                return rm_node.data;
            }

            this.head = rm_node.next;

            rm_node.next = undefined;

            return rm_node.data;
        }

        this.length = Math.max(0, --this.length);

        let previous_node: TNode<T> | undefined = undefined;
        let current_node: TNode<T> | undefined = this.head;

        for (let i = 0; i < idx && current_node; ++i) {
            previous_node = current_node;
            current_node = current_node.next;
        }

        if (this.length === 0) {
            const rm_node = current_node;

            this.head = this.tail = undefined;

            return rm_node?.data;
        }

        // Handle tail removal
        if (current_node && !current_node.next) {
            const rm_node = current_node;

            previous_node!.next = undefined;

            this.tail = previous_node;

            rm_node.next = undefined;

            return rm_node.data;
        }

        const rm_node = current_node;

        previous_node!.next = current_node?.next;

        rm_node!.next = undefined;

        return rm_node?.data;
    }

    get(idx: number): T | undefined {
        if (this.length === 0) {
            return;
        }

        if (idx < 0 || idx > this.length) {
            throw new Error("index out of bounds");
        }

        if (this.head && idx === 0) {
            return this.head.data;
        }

        if (this.tail && idx === this.length) {
            return this.tail.data;
        }

        let previous_node: TNode<T> | undefined = undefined;
        let current_node: TNode<T> | undefined = this.head;

        for (let i = 0; i < idx && current_node; ++i) {
            previous_node = current_node;
            current_node = current_node.next;
        }

        return current_node?.data;
    }

    print(): void {
        let msg = "";

        msg += `Len: ${this.length} `;

        let current_node: TNode<T> | undefined = this.head;

        while (current_node) {
            msg += `[${current_node.data}] -> `;
            current_node = current_node.next;
        }

        msg += "END";
        console.log(msg);
        console.log("HEAD:", this.head ? this.head.data : "undefined");
        console.log("TAIL:", this.tail ? this.tail.data : "undefined");
    }
}

const linked_list = new LinkedList();

linked_list.print();

linked_list.insertAt(10, 0);

linked_list.print();

linked_list.append(20);

linked_list.print();

linked_list.prepend(30);

linked_list.print();

linked_list.insertAt(40, 1);

linked_list.print();

linked_list.insertAt(50, linked_list.length - 1);

linked_list.print();

console.log("Removed:", linked_list.remove(20));

linked_list.print();

console.log("Removed:", linked_list.removeAt(1));

linked_list.print();

console.log("Removed:", linked_list.removeAt(linked_list.length - 1));

linked_list.print();

console.log("Removed:", linked_list.removeAt(0));

linked_list.print();

linked_list.append(30);

linked_list.print();

linked_list.append(40);

linked_list.print();

console.log("Removed:", linked_list.removeAt(0));

linked_list.print();

for (let i = linked_list.length - 1; i >= 0; --i) {
    console.log("Removed:", linked_list.remove(linked_list.get(i)));
    linked_list.print();
}

for (let i = 0; i < 6; ++i) {
    linked_list.insertAt(Math.floor(Math.random() * 100), i);
    linked_list.print();
}

for (let i = 0; i < Math.floor(linked_list.length / 2); ++i) {
    console.log(
        "Removed:",
        linked_list.removeAt(Math.floor(Math.random() * linked_list.length)),
    );
    linked_list.print();
}

File: c/singly_linked_list.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

typedef struct LinkedList {
    size_t length;
    struct Node *head;
    struct Node *tail;
} LinkedList;

void prepend(LinkedList *list, int data) {
    Node *node = (Node *)malloc(sizeof(Node));

    if (node == NULL) {
        fprintf(stderr, "ERROR: malloc failed in prepend\n");
        exit(1);
    }

    node->data = data;
    node->next = NULL;

    if (list->length == 0) {
        list->head = node;
        list->tail = node;
    } else {
        node->next = list->head;
        list->head = node;
    }

    list->length++;
}

void append(LinkedList *list, int data) {
    Node *node = (Node *)malloc(sizeof(Node));

    if (node == NULL) {
        fprintf(stderr, "ERROR: malloc failed in append\n");
        exit(1);
    }

    node->data = data;
    node->next = NULL;

    if (list->length == 0) {
        list->head = node;
        list->tail = node;
    } else {
        list->tail->next = node;
        list->tail = node;
    }

    list->length++;
}

void insert_at(LinkedList *list, int data, size_t idx) {
    if (idx > list->length) {
        fprintf(stderr, "ERROR: index out of bounds");
        return;
    }

    if (idx == 0) {
        prepend(list, data);
        return;
    }

    if (idx == list->length) {
        append(list, data);
        return;
    }

    list->length++;

    Node *node = (Node *)malloc(sizeof(Node));

    if (node == NULL) {
        fprintf(stderr, "ERROR: malloc failed in insert_at\n");
        exit(1);
    }

    node->data = data;
    node->next = NULL;

    Node *previous_node = NULL;
    Node *current_node = list->head;

    for (size_t i = 0; i < idx && current_node != NULL; ++i) {
        previous_node = current_node;
        current_node = current_node->next;
    }

    previous_node->next = node;
    node->next = current_node;
}

int remove_node(LinkedList *list, int data) {
    if (list == NULL || list->length == 0) {
        return -1;
    }

    if (list->head->data == data) {
        list->length--;

        Node *rm_node = list->head;

        if (list->length == 0) {
            list->head = NULL;
            list->tail = NULL;
        } else {
            list->head = list->head->next;
        }

        int rm_data = rm_node->data;

        free(rm_node);

        return rm_data;
    }

    Node *previous_node = NULL;
    Node *current_node = list->head;

    while (current_node != NULL) {
        if (current_node->data == data) {
            list->length--;

            if (current_node->next == NULL) {
                Node *rm_node = list->tail;

                list->tail = previous_node;

                int rm_data = rm_node->data;

                free(rm_node);

                return rm_data;
            }

            Node *rm_node = current_node;

            previous_node->next = current_node->next;

            int rm_data = rm_node->data;

            free(rm_node);

            return rm_data;
        }

        previous_node = current_node;
        current_node = current_node->next;
    }

    return -1;
}

int remove_node_at(LinkedList *list, size_t idx) {
    if (list == NULL || list->length == 0) {
        return -1;
    }

    if (idx > list->length) {
        fprintf(stderr, "ERROR: index out of bounds");
        return -1;
    }

    list->length--;

    if (idx == 0) {
        Node *rm_node = list->head;

        if (list->length == 0) {
            list->head = NULL;
            list->tail = NULL;
        } else {
            list->head = list->head->next;
        }

        int rm_data = rm_node->data;

        free(rm_node);

        return rm_data;
    }

    Node *previous_node = NULL;
    Node *current_node = list->head;

    for (size_t i = 0; i < idx && current_node != NULL; ++i) {
        previous_node = current_node;
        current_node = current_node->next;
    }

    if (current_node->next == NULL) {
        Node *rm_node = list->tail;

        list->tail = previous_node;

        int rm_data = rm_node->data;

        free(rm_node);

        return rm_data;
    }

    Node *rm_node = current_node;

    previous_node->next = current_node->next;

    int rm_data = rm_node->data;

    free(rm_node);

    return rm_data;
}

int get(LinkedList *list, size_t idx) {
    if (list == NULL || list->length == 0) {
        return -1;
    }

    if (idx > list->length) {
        fprintf(stderr, "ERROR: index out of bounds");
        return -1;
    }

    if (idx == 0) {
        return list->head->data;
    }

    if (idx == list->length) {
        return list->tail->data;
    }

    Node *current_node = list->head;

    for (size_t i = 0; i < idx && current_node != NULL; ++i) {
        current_node = current_node->next;
    }

    return current_node->data;
}

void print_list(LinkedList *list) {
    char *msg = (char *)malloc(sizeof(char) * 1024);

    if (msg == NULL) {
        fprintf(stderr, "ERROR: malloc failed in print\n");
        exit(1);
    }

    char *ptr = msg;

    ptr += sprintf(ptr, "Length: %zu ", list->length);

    Node *current_head = list->head;

    while (current_head != NULL) {
        ptr += sprintf(ptr, "[%d] -> ", current_head->data);
        current_head = current_head->next;
    }

    sprintf(ptr, "END");

    printf("%s\n", msg);

    free(msg);
}

int main(void) {
    LinkedList linked_list = {0, NULL, NULL};

    print_list(&linked_list);

    insert_at(&linked_list, 5, 0);
    append(&linked_list, 7);
    append(&linked_list, 9);

    print_list(&linked_list);

    assert(get(&linked_list, 2) == 9);
    assert(remove_node_at(&linked_list, 1) == 7);
    print_list(&linked_list);
    assert(linked_list.length == 2);

    print_list(&linked_list);

    append(&linked_list, 11);

    assert(remove_node_at(&linked_list, 1) == 9);
    assert(remove_node(&linked_list, 9) == -1);
    assert(remove_node_at(&linked_list, 0) == 5);
    assert(remove_node_at(&linked_list, 0) == 11);
    assert(linked_list.length == 0);

    print_list(&linked_list);

    prepend(&linked_list, 5);
    prepend(&linked_list, 7);
    prepend(&linked_list, 9);

    print_list(&linked_list);

    assert(get(&linked_list, 2) == 5);
    assert(get(&linked_list, 0) == 9);
    assert(remove_node(&linked_list, 9) == 9);
    assert(linked_list.length == 2);
    assert(get(&linked_list, 0) == 7);

    printf("ALL TESTS PASSED\n");

    return 0;
}
